### PR TITLE
Better completion callback for activate

### LINKED ISSUE #3586: "Most recently fetched config is already activated" after remote config activate() call
## [REQUIRED] Step 2: Describe your environment

  * Xcode version: 10.3
  * Firebase SDK version: 6.6.0
  * Firebase Component: Remote Config
  * Component version: 4.2.1

### [REQUIRED] Step 3: Describe the problem
We fetch and activate remote config with the following code:
```
let expirationDuration: TimeInterval
#if DEBUG
expirationDuration = 0
#else
expirationDuration = 3600
#endif
firebaseRemoteConfig.fetch(withExpirationDuration: expirationDuration) { [unowned self] (status, error) in
    if status == .success {
        log.info("Remote Config fetched")
        self.firebaseRemoteConfig.activate(completionHandler: { (error) in
            if let error = error {
                log.error(error)
            }
        })
    } else {
        log.error("Remote Config could not be fetched. \(String(describing: error))")
    }
}
```

We call this part of the code only one, in `AppDelegate` during `didFinishLaunchingWithOptions` 

In the wild, we see lots of Remote config errors reported, they're pretty straightforward:

> ActivationFailureReason: Most recently fetched config is already activated
> Code: 8003

Is it an error to try to activate fetched config when it's the same as active one? If yes, which approach should we take?
We can't use the `fetchAndActivate` since it doesn't have the `expirationDuration` parameter.

### PR DESCRIPTION
**Pull request by paulb777**
Adds an activate API with a changed parameter in the callback. It replaces a deprecated API that set an internal error if there was no change.

Fixes #3586 

The integration test infrastructure is implemented in #5633 

Googlers: See accompanying API proposal.

### CODE REVIEW

File: FirebaseRemoteConfig/Sources/Public/FIRRemoteConfig.h [hunk 1]
```diff
@@ -232,18 +238,23 @@ NS_SWIFT_NAME(RemoteConfig)
...
+/// of the app to take effect (depending on how config data is used in the app).
+/// @param completion Activate operation callback.
+- (void)activateWithCompletion:(nullable FIRRemoteConfigActivateChangeCompletion)completion;
+
 /// Applies Fetched Config data to the Active Config, causing updates to the behavior and appearance
 /// of the app to take effect (depending on how config data is used in the app).
 /// @param completionHandler Activate operation callback.
-- (void)activateWithCompletionHandler:(nullable FIRRemoteConfigActivateCompletion)completionHandler;
+- (void)activateWithCompletionHandler:(nullable FIRRemoteConfigActivateCompletion)completionHandler
+    DEPRECATED_MSG_ATTRIBUTE("Use -[FIRRemoteConfig activateWithHandler:] instead.");
```
Comment (paulb777) [inline FirebaseRemoteConfig/Sources/Public/FIRRemoteConfig.h:250]: Swift friendly message.

File: FirebaseRemoteConfig/Sources/FIRRemoteConfig.m [hunk 2]
```diff
@@ -278,37 +278,54 @@ - (BOOL)activateFetched {
...
   __weak FIRRemoteConfig *weakSelf = self;
   void (^applyBlock)(void) = ^(void) {
     FIRRemoteConfig *strongSelf = weakSelf;
     if (!strongSelf) {
       NSError *error = [NSError errorWithDomain:FIRRemoteConfigErrorDomain
                                            code:FIRRemoteConfigErrorInternalError
                                        userInfo:@{@"ActivationFailureReason" : @"Internal Error."}];
-      if (completionHandler) {
+      if (completion) {
         dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
```
Comment (chliangGoogle) [inline FirebaseRemoteConfig/Sources/FIRRemoteConfig.m:299]: callback should be called in main queue, otherwise might block UI. (altho seems not related to this PR, although seems like it's using a bad practice for all following places)

File: FirebaseRemoteConfig/Sources/Public/FIRRemoteConfig.h [hunk 3]
```diff
@@ -87,6 +87,12 @@ typedef void (^FIRRemoteConfigFetchCompletion)(FIRRemoteConfigFetchStatus status
 typedef void (^FIRRemoteConfigActivateCompletion)(NSError *_Nullable error)
     NS_SWIFT_NAME(RemoteConfigActivateCompletion);
 
+/// Completion handler invoked by activate method upon completion.
+/// @param changed  Flag indicating whether or not the configuration changed.
+/// @param error  Error message on failure. Nil if activation was successful.
+typedef void (^FIRRemoteConfigActivateChangeCompletion)(BOOL changed, NSError *_Nullable error)
```
Comment (chliangGoogle) [inline FirebaseRemoteConfig/Sources/Public/FIRRemoteConfig.h:93]: can you illustrate a bit on what the advantage is. I'm guessing using the boolean as first parameter will improve checking in swift but I don't know what exactly it is.

File: FirebaseRemoteConfig/Tests/HermeticAPI/APITests.swift [hunk 4]
```diff
@@ -130,6 +125,41 @@ class APITests: XCTestCase {
     waitForExpectations()
   }
 
+  // Test New API.
+  func testUnchangedActivateWillFlag() {
+    let expectation = self.expectation(description: #function)
+    config.fetch { status, error in
```
Comment (chliangGoogle) [inline FirebaseRemoteConfig/Tests/HermeticAPI/APITests.swift:131]: Will this also help us adding ResultType?

File: FirebaseRemoteConfig/Sources/FIRRemoteConfig.m (hunk 2 above)
Comment (paulb777) [reply FirebaseRemoteConfig/Sources/FIRRemoteConfig.m:299]: Probably should be a separate PR, right?

File: FirebaseRemoteConfig/Sources/Public/FIRRemoteConfig.h (hunk 3 above)
Comment (paulb777) [reply FirebaseRemoteConfig/Sources/Public/FIRRemoteConfig.h:93]: It gives feedback to the API client about whether or not the config changed without an error being thrown. See #3586 for more on the issues caused by the other API.

File: FirebaseRemoteConfig/Tests/HermeticAPI/APITests.swift (hunk 4 above)
Comment (paulb777) [reply FirebaseRemoteConfig/Tests/HermeticAPI/APITests.swift:131]: Yep, but adding a Swift extension for RemoteConfig is out of scope for now.

File: FirebaseRemoteConfig/Sources/Public/FIRRemoteConfig.h (hunk 3 above)
Comment (chliangGoogle) [reply FirebaseRemoteConfig/Sources/Public/FIRRemoteConfig.h:93]: Nice, so this will align with Android.

Comment (paulb777) [reply FirebaseRemoteConfig/Sources/Public/FIRRemoteConfig.h:93]: And JS :)

File: FirebaseRemoteConfig/Tests/SwiftAPI/APITests.swift [hunk 5]
```diff
@@ -110,6 +105,41 @@ class APITests: APITestBase {
...
+    let expectation = self.expectation(description: #function)
+    config.fetch { status, error in
+      if let error = error {
+        XCTFail("Fetch Error \(error)")
+      }
+      XCTAssertEqual(status, RemoteConfigFetchStatus.success)
+      self.config.activate { changed, error in
+        XCTAssertTrue(!APITests.useFakeConfig || changed)
+        if let error = error {
+          print("Activate Error \(error)")
```
Comment (karenyz) [inline FirebaseRemoteConfig/Tests/SwiftAPI/APITests.swift:120]: should this use `XCTFail` like it does above?

File: FirebaseRemoteConfig/Sources/FIRRemoteConfig.m [hunk 6]
```diff
@@ -278,37 +278,56 @@ - (BOOL)activateFetched {
...
       FIRLogError(kFIRLoggerRemoteConfig, @"I-RCN000068", @"Internal error activating config.");
       return;
     }
     // Check if the last fetched config has already been activated. Fetches with no data change are
     // ignored.
+    NSLog(@"%f %f", strongSelf->_settings.lastETagUpdateTime,
+          strongSelf->_settings.lastApplyTimeInterval);
     if (strongSelf->_settings.lastETagUpdateTime == 0 ||
         strongSelf->_settings.lastETagUpdateTime <= strongSelf->_settings.lastApplyTimeInterval) {
       FIRLogWarning(kFIRLoggerRemoteConfig, @"I-RCN000069",
```
Comment (karenyz) [inline FirebaseRemoteConfig/Sources/FIRRemoteConfig.m:316]: since we're no longer treating this as an error I wonder if we should downgrade the log severity

File: FirebaseRemoteConfig/Tests/SwiftAPI/APITests.swift (hunk 5 above)
Comment (paulb777) [reply FirebaseRemoteConfig/Tests/SwiftAPI/APITests.swift]: This new API should assert on error being nil. Good catch!

File: FirebaseRemoteConfig/Sources/FIRRemoteConfig.m (hunk 6 above)
Comment (paulb777) [reply FirebaseRemoteConfig/Sources/FIRRemoteConfig.m]: Done. Thanks!

File: FirebaseRemoteConfig/Tests/FakeConsole/FakeConsoleTests.swift [hunk 7]
```diff
@@ -59,6 +60,43 @@ class FakeConsoleTests: APITestBase {
...
+  func testChangedActivateWillNotFlag() {
+    let expectation = self.expectation(description: #function)
+    config.fetch { status, error in
+      if let error = error {
+        XCTFail("Fetch Error \(error)")
+      }
+      XCTAssertEqual(status, RemoteConfigFetchStatus.success)
+      self.config.activate { changed, error in
+        XCTAssertNil(error)
+        XCTAssert(changed)
```
Comment (ryanwilson) [inline FirebaseRemoteConfig/Tests/FakeConsole/FakeConsoleTests.swift:74]: Tiny nit: Can you change this to `XCTAssertTrue`? I did a double take on it, I think explicitly saying `True` makes it more clear.

File: FirebaseRemoteConfig/Tests/FakeConsole/FakeConsoleTests.swift [hunk 8]
```diff
@@ -59,6 +60,43 @@ class FakeConsoleTests: APITestBase {
     waitForExpectations()
   }
 
+  // Test New API.
+  // Contrast with testUnchangedActivateWillFlag in APITests.swift.
+  func testChangedActivateWillNotFlag() {
```
Comment (ryanwilson) [inline FirebaseRemoteConfig/Tests/FakeConsole/FakeConsoleTests.swift:65]: Should this also test the `false, nil` case where activate is called but no changes happened?

File: FirebaseRemoteConfig/Tests/FakeConsole/FakeConsoleTests.swift (hunk 7 above)
Comment (paulb777) [reply FirebaseRemoteConfig/Tests/FakeConsole/FakeConsoleTests.swift]: Agreed! Done

File: FirebaseRemoteConfig/Tests/FakeConsole/FakeConsoleTests.swift (hunk 8 above)
Comment (paulb777) [reply FirebaseRemoteConfig/Tests/FakeConsole/FakeConsoleTests.swift:65]: Yes, but that test doesn't need the FakeConsole and was added to APITests.swift
